src/plugin-traverser.ts

```diff
@@ -0,0 +1,115 @@
+import * as bluebirdPromise from 'bluebird';
+import * as fs from 'fs';
+import * as path from 'path';
+
+import { BasePlugin } from './abstract/base-plugin.plugin';
+import { PluginModuleOptions } from './interfaces/plugin-module-options.interface';
+import { PluginOptions } from './interfaces/plugin-options.interface';
+import { Provider } from '@nestjs/common';
+
+const node_modules = require('node_modules-path');
+
+export class PluginTraverser {
+  private _pluginType: string;
+  private _directories: Array<string> = [path.resolve(node_modules())];
+
+  constructor(pluginModuleOptions: PluginModuleOptions) {
+    this._pluginType = pluginModuleOptions.pluginType;
+    if (
+      pluginModuleOptions.directories &&
+      pluginModuleOptions.directories.length > 0
+    )
+      this._directories = pluginModuleOptions.directories;
+  }
+
+  public async traverseDirectories(): Promise<Array<Provider<BasePlugin>>> {
+    const modules = [];
+    await bluebirdPromise.each(this._directories, (parentDirectory) => {
+      if (fs.existsSync(parentDirectory)) {
+        fs.readdirSync(parentDirectory, { withFileTypes: true })
```

Comment:
We are still reading synchronously, checkin existence synchronously even we use promise.each. We should complete file operations asynchronously.

fs.exists is deprecated by the way, so you should find another alternative like fs.open, fs.stat.

Reply:
We cannot perform async operation in the method due to the need for output.